src/main/java/io/temporal/internal/statemachines/SignalExternalStateMachine.java

```diff
@@ -0,0 +1,172 @@
+
+  /**
+   * Register new instance of the signal commands
+   *
+   * @param signalAttributes attributes used to signal an external workflow
+   * @param completionCallback either SignalExternalWorkflowExecutionFailed,
+   *     ExternalWorkflowExecutionSignaled or true value of the second parameter to indicate
+   *     immediate cancellation.
+   * @param commandSink sink to send commands
+   * @return cancellation handler
+   */
+  public static Functions.Proc newInstance(
+      SignalExternalWorkflowExecutionCommandAttributes signalAttributes,
+      Functions.Proc2<Void, Failure> completionCallback,
+      Functions.Proc1<CancellableCommand> commandSink,
+      Functions.Proc1<StateMachine> stateMachineSink) {
+    SignalExternalStateMachine commands =
+        new SignalExternalStateMachine(
+            signalAttributes, completionCallback, commandSink, stateMachineSink);
+    return commands::cancel;
+  }
+
+  private SignalExternalStateMachine(
+      SignalExternalWorkflowExecutionCommandAttributes signalAttributes,
+      Functions.Proc2<Void, Failure> completionCallback,
+      Functions.Proc1<CancellableCommand> commandSink,
+      Functions.Proc1<StateMachine> stateMachineSink) {
+    super(STATE_MACHINE_DEFINITION, commandSink, stateMachineSink);
+    this.signalAttributes = signalAttributes;
+    this.completionCallback = completionCallback;
+    explicitEvent(ExplicitEvent.SCHEDULE);
+  }
+
+  enum ExplicitEvent {
+    SCHEDULE,
+    CANCEL
+  }
+
+  enum State {
+    CREATED,
+    SIGNAL_EXTERNAL_COMMAND_CREATED,
+    SIGNAL_EXTERNAL_COMMAND_RECORDED,
+    SIGNALED,
+    FAILED,
+    CANCELED,
+  }
+
+  public static final StateMachineDefinition<State, ExplicitEvent, SignalExternalStateMachine>
+      STATE_MACHINE_DEFINITION =
+          StateMachineDefinition.<State, ExplicitEvent, SignalExternalStateMachine>newInstance(
+                  "SignalExternal", State.CREATED, State.SIGNALED, State.FAILED, State.CANCELED)
+              .add(
+                  State.CREATED,
+                  ExplicitEvent.SCHEDULE,
+                  State.SIGNAL_EXTERNAL_COMMAND_CREATED,
+                  SignalExternalStateMachine::createSignalExternalCommand)
+              .add(
+                  State.SIGNAL_EXTERNAL_COMMAND_CREATED,
+                  CommandType.COMMAND_TYPE_SIGNAL_EXTERNAL_WORKFLOW_EXECUTION,
+                  State.SIGNAL_EXTERNAL_COMMAND_CREATED)
+              .add(
+                  State.SIGNAL_EXTERNAL_COMMAND_CREATED,
+                  ExplicitEvent.CANCEL,
+                  State.CANCELED,
+                  SignalExternalStateMachine::cancelSignalExternalCommand)
+              .add(
+                  State.SIGNAL_EXTERNAL_COMMAND_CREATED,
+                  EventType.EVENT_TYPE_SIGNAL_EXTERNAL_WORKFLOW_EXECUTION_INITIATED,
+                  State.SIGNAL_EXTERNAL_COMMAND_RECORDED,
+                  EntityStateMachineInitialCommand::setInitialCommandEventId)
+              .add(
```

Comment:
basically, once we have a notification from the server that the external workflow was signaled, then a cancel is basically a no-op, hence this transition back to its own state?

Reply:
Yes, as there is no cancellation support by the service.